### PR TITLE
fix(types): Origin within OriginFunction can be undefined

### PR DESCRIPTION
**Pull request by joshmeads**
In reference to #236 origin cannot be assumed to always be a string. Propose making it `string | undefined` to ensure people handle it correctly. This would flag quickly in the sync example on the README that `new URL` constructor can't work with undefined and should be tested to ensure it's a string first.

This also matches an existing [test setup:](https://github.com/fastify/fastify-cors/blob/1b8950ec2309f265a9861d3c52bd2498c4826569/test/index.test-d.ts#L91-L97)
```ts
const corsDelegate: OriginFunction = (origin, cb) => {
  if (typeof origin === 'undefined' || /localhost/.test(origin)) {
    cb(null, true)
    return
  }
  cb(new Error(), false)
}
```

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by Uzlopak** [comment]
Can you add a typing test please?

File: index.d.ts [hunk 1]
```diff
@@ -22,7 +22,7 @@ type FastifyCorsHook =
   | 'onSend'
 
 declare namespace fastifyCors {
-  export type OriginFunction = (origin: string, callback: OriginCallback) => void;
+  export type OriginFunction = (origin?: string, callback: OriginCallback) => void;
```
Comment (climba03003) [inline index.d.ts:25]: Do not use `?:` when the later argument is not optional.

```suggestion
  export type OriginFunction = (origin: string | undefined, callback: OriginCallback) => void;
```

**Review by climba03003 — CHANGES_REQUESTED** [review]

File: index.d.ts (hunk 1 above)
Comment (joshmeads) [reply index.d.ts]: I wrote that first and then removed it to a `?` after seeing no other examples of `| undefined` used. Fully agree with this though! Thanks!

**Comment by joshmeads** [comment]
> Can you add a typing test please?

@Uzlopak apologies for the slow reply, busy time of year. What do you mean by typing test exactly? The altered code matches the test within `tests/index.test-d.ts` linked in the description. I can't see any type specific tests within the codebase unless I'm missing something obvious. Happy to do it though - just need a pointer on where to look :) 

Thanks guys!

**Comment by Uzlopak** [comment]
Well, what issue does your PR fix. As you didnt change any test, we have to assume nothing was wrong in the first place...

**Comment by jeanvalentin** [comment]
This PR would prevent users from blindly calling `new URL(origin)` and throwing, like I just did. I assumed `origin` was always a valid URL; it is in fact not even guaranteed to be a string. Good call. You might want to update the example in the readme as well, by adding an instruction such as `if (!origin) return cb(null, true);` on top.

Another option is to prevent CORS checks from being performed when `origin` is empty, since that would not be cross-origin.

**Comment by Uzlopak** [comment]
How about patching tests/index.test-d.ts to reflect that?

**Comment by Uzlopak** [comment]
patched accordingly

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**